subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyCollector.java

```diff
@@ -166,6 +178,59 @@ public <D extends Dependency> void add(Provider<? extends D> dependency, Action<
         doAddLazy(dependency, configuration);
     }
 
+
+    private static DependencyConstraint applyConstraintConfiguration(DependencyConstraint dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        if (config != null) {
+            config.execute(dependencyConstraint);
+        }
+
+        // TODO DependencyConstraint mutation validator?
```

Comment:
🤔 Do we not check for mutations in the top-level block today?

Reply:
I can't say I'm not surprised we don't implement mutation checks for constraints. The more you look, the more mutation checks you find we're missing. 

I think we should fix this, but separately. We should write up an issue for this and put it on the board

Reply:
https://github.com/gradle/gradle/issues/27904